frontend/src/globals/unity-indexeddb-worker.ts

```diff
@@ -0,0 +1,359 @@
+				// If we end up with multiple parallel read transactions for a
+				// while because we've cleared this reference, that is no
+				// problem.
+
+				readTransaction.oncomplete = () => {
+					this.numReadTransactions--;
+					this.commitMemoryCache();
+				};
+
+				readTransaction.onerror = () => {
+					this.numReadTransactions--;
+				};
+
+				readTransaction.onabort = () => {
+					this.numReadTransactions--;
+				};
+
+				this.numReadTransactions++;
+			} catch (e) {
+				if (e.name === 'InvalidStateError' && this.db !== undefined) {
+					this.recreateIndexedDb();
+				}
+
+				// Requests aren't queued anywhere, so if one fails it won't
+				// automatically try again, so signal the viewer that the request
+				// has been handled to prevent stalling.
+
+				this.sendGetTransactionComplete({
+					id,
+					size: -1,
+				});
+
+				console.error('Unexpected IndexedDb read exception', e);
+			}
+		}
+	}
+
+	sendSetTransactionComplete(parms: any) {
+		self.postMessage({
+			type: 'OnSetTransactionComplete',
+			parms,
+		});
+	}
+
+	sendGetTransactionComplete(parms: any) {
+		self.postMessage({
+			type: 'OnGetTransactionComplete',
+			parms,
+		});
+	}
+
+	sendIndexedDbUpdated(parms: any) {
+		self.postMessage({
+			type: 'OnIndexedDbUpdated',
+			parms,
+		});
+	}
+
+	commitMemoryCache() {
+		try {
+			// This function commits records in memory to indexeddb. Only when
+			// records have been successfully written are they removed from
+			// the memory cache.
+			// If something goes wrong (e.g. the browser tab is closed), the
+			// memory cache is lost, which is not ideal, but also not terrible
+			// for our use case.
+
+			// If we already have a readwrite transaction running, then don't
+			// do anything. readwrite transactions cannot overlap, and this
+			// function will chain them when there is nothing else to do
+			// automatically. readwrites will also block reads, which are more
+			// important to the user experience, so dont attempt to write if
+			// there are any outstanding (when they are complete, they will call
+			// back here to try again.)
+
+			if (this.committing) {
+				return;
+			}
+			if (this.numReadTransactions > 0) {
+				return;
+			}
+			const keys = Object.keys(this.memoryCache);
+			if (keys.length <= 0) {
+				return;
+			}
+
+			this.committing = true;
+
+			const transaction = this.db.transaction(this.objectStoreName, 'readwrite');
+			const objectStore = transaction.objectStore(this.objectStoreName);
+
+			let numRequests = 0;
+
+			// eslint-disable-next-line guard-for-in
+			for (const key of keys) {
+				const record = this.memoryCache[key];
+				const request = objectStore.put(record, key);
+				request.onsuccess = () => {
+					delete this.memoryCache[key]; // Delete the local copy
+					this.index[key] = 1; // Update the index as we go
+				};
+				if (numRequests++ > 5) { // Don't try and commit too much in one transaction otherwise we may block users' get requests
```

Comment:
it looks like every iteration would create a request? wouldn't it be better to just loop through `keys.slice(0, 5)` instead of doing a break?

then you wouldn't need to declare numRequests either

Reply:
Yes, that is much neater.